metrics/config.go

```diff
@@ -0,0 +1,127 @@
+)
+
+type metricsConfig struct {
+	// The metrics domain. e.g. "serving.knative.dev" or "build.knative.dev".
+	domain string
+	// The component that emits the metrics. e.g. "activator", "autoscaler".
+	component string
+	// The metrics backend destination.
+	backendDestination MetricsBackend
+	// The stackdriver project ID where the stats data are uploaded to. This is
+	// not the GCP project ID.
+	stackdriverProjectID string
+}
+
+var domainRE = regexp.MustCompile("^(serving|build|eventing)\\.knative\\.dev$")
+
+func getMetricsConfig(m map[string]string, domain string, component string, logger *zap.SugaredLogger) (*metricsConfig, error) {
+	var mc metricsConfig
+	backend, ok := m[backendDestinationKey]
+	if !ok {
+		return nil, errors.New("metrics.backend-destination key is missing")
+	}
+	lb := MetricsBackend(strings.ToLower(backend))
+	switch lb {
+	case Stackdriver, Prometheus:
+		mc.backendDestination = lb
+	default:
+		return nil, fmt.Errorf("Unsupported metrics backend value \"%s\"", backend)
+	}
+
+	if mc.backendDestination == Stackdriver {
+		sdProj, ok := m[stackdriverProjectIDKey]
+		if !ok || sdProj == "" {
+			return nil, errors.New("For backend stackdriver, metrics.stackdriver-project-id field must exist and cannot be empty")
+		}
+		mc.stackdriverProjectID = sdProj
+	}
+
+	if !domainRE.MatchString(domain) {
+		return nil, fmt.Errorf("Invalid metrics domain name \"%s\"", domain)
+	}
+	mc.domain = domain
+
+	if component == "" {
+		return nil, errors.New("Metrics component name cannot be empty")
+	}
+	mc.component = component
+	return &mc, nil
+}
+
+// UpdateExporterFromConfigMap returns a helper func that can be used to update the exporter
+// when a config map is updated
+func UpdateExporterFromConfigMap(domain string, component string, logger *zap.SugaredLogger) func(configMap *corev1.ConfigMap) {
+	return func(configMap *corev1.ConfigMap) {
+		newConfig, err := getMetricsConfig(configMap.Data, domain, component, logger)
+		if err != nil {
+			ce := getCurMetricsExporter()
+			if ce == nil {
+				// Fail the process if there doesn't exist an exporter.
+				logger.Fatal("Failed to get a valid metrics config")
+			} else {
+				logger.Error("Failed to get a valid metrics config; Skip updating the metrics exporter", zap.Error(err))
+				return
+			}
+		}
+
+		if isMetricsConfigChanged(newConfig) {
+			if err := newMetricsExporter(newConfig, logger); err != nil {
+				logger.Error("Failed to update a new metrics exporter based on metric config.", zap.Error(err))
+				return
+			}
+		}
+	}
+}
+
+// isMetricsConfigChanged compares the non-nil newConfig against curMetricsConfig. When backend changes,
+// or stackdriver project ID changes for stackdriver backend, we need to update the metrics exporter.
+func isMetricsConfigChanged(newConfig *metricsConfig) bool {
+	changed := false
+	cc := getCurMetricsConfig()
+	if cc == nil || newConfig.backendDestination != cc.backendDestination {
+		changed = true
```

Comment:
nit: you can remove changed variable and short circuit trues with a return. Example:
```
	cc := getCurMetricsConfig()
	if cc == nil || newConfig.backendDestination != cc.backendDestination {
		return true
	} else if newConfig.backendDestination == Stackdriver && newConfig.stackdriverProjectID != cc.stackdriverProjectID {
		return true
	}
	return false
```